lib/drivers/docker/index.js

```diff
@@ -0,0 +1,380 @@
+
+  if (opts.size != null && opts.size !== original.size) {
+    debug('cluster size changed, sending set-size: %j => %j', id, opts);
+    this.requestOfInstance(id, {cmd: 'set-size', size: opts.size});
+  }
+};
+
+DockerDriver.prototype.removeInstance = function(id, cb) {
+  debug('removeInstance(%j)', id);
+  var instance = this._instance(id);
+  if (instance.current) {
+    instance.current.kill('SIGKILL', function(err) {
+      instance.current.destroy();
+      delete instance.current;
+      cb(err);
+    });
+  } else {
+    setImmediate(cb);
+  }
+};
+
+DockerDriver.prototype.deployInstance = function(id, req, res) {
+  debug('deployInstance(%j)', id);
+  var self = this;
+  var instance = this._instance(id);
+  var image = this.Image.from(this, id, this.baseDir, req, res);
+  instance.nextImage = image;
+  image.on('error', function(err) {
+    self.error('error deploying: %s, %j', err, err);
+    throw err;
+  });
+  image.on('image', function() {
+    debug('Deployment image:', image.name);
+    if (instance.current) {
+      instance.current.destroy(start);
+    } else {
+      start();
+    }
+  });
+
+  function start() {
+    self.startInstance(id, function(err) {
+      if (err) {
+        self.error('Error deploying instance %j:', id, err);
+      } else {
+        self.log('Instance %j deployed', id);
+      }
+    });
+  }
+};
+
+DockerDriver.prototype.startInstance = function(id, cb) {
+  debug('startInstance(%j, %j)', id, cb.name);
+  var self = this;
+  var instance = this._instance(id);
+  var imgToLaunch = instance.nextImage ||
+                    instance.currentImage ||
+                    instance.previousImage;
+  this.server.getInstanceEnv(id, function(err, env) {
+    if (err) return cb(err);
+    var opts = _.merge(instance.startOpts, {env: env});
+    instance.log.enableGC();
+    instance.ports = [];
+    instance.previousImage = instance.currentImage;
+    instance.currentImage = imgToLaunch;
+    delete instance.nextImage;
+    instance.current = imgToLaunch.start(id, instance.log, opts);
+    instance.current.on('request', function(req) {
+      debug('bubbling up request from instance %j:', id,
+            _.pick(req, ['pid', 'wid', 'pst', 'cmd']));
+      self.emit('request', id, req);
+    });
+    instance.current.on('error', function(err) {
+      self.error('error from docker child %j:', id, err);
+    });
+    instance.current.on('exit', function(err) {
+      if (isError(err) && err.statusCode === 404 && this === instance.current) {
+        self.startInstance(id, function() {
+          self.log('restarted instance after container disappeared (exit)');
+        });
+      }
+    });
+    if (cb) {
+      instance.current.on('connected', cb);
+    }
+  });
+};
+
+DockerDriver.prototype._onRequest = function(instanceId, req) {
+  var instance = this._instance(instanceId);
+  switch (req.cmd) {
+    case 'listening':
+      // emit a listening event each time a _new_ port is listened on
```

Comment:
I don't think the direct driver does this filtering, is it necessary to not multiply emit?